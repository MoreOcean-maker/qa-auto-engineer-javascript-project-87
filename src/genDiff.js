import getParsedFile from './getParsedFile.js';
import buildDiff from './buildDiff.js';

const stringifyPlain = (value) => {
  if (value === null) return 'null';
  if (typeof value === 'object') return '[complex value]';
  return String(value);
};

const stringifyStylish = (value, depth) => {
  if (typeof value === 'object' && value !== null) {
    const indent = ' '.repeat(4 * (depth + 1));
    const entries = Object.entries(value);
    return `{\n${entries.map(([k, v]) => `${indent}${k}: ${stringifyStylish(v, depth + 1)}`).join('\n')}\n${' '.repeat(4 * depth)}}`;
  }
  return String(value);
};

const formatters = {
  plain: (diff) => {
    const iter = (nodes, parentPath = '') => nodes
      .flatMap((node) => {
        const currentPath = parentPath ? `${parentPath}.${node.key}` : node.key;

        switch (node.type) {
          case 'added':
            return `Property '${currentPath}' was added with value: ${stringifyPlain(node.value)}`;
          case 'removed':
            return `Property '${currentPath}' was removed`;
          case 'updated':
            return `Property '${currentPath}' was updated. From ${stringifyPlain(node.oldValue)} to ${stringifyPlain(node.value)}`;
          case 'nested':
            return iter(node.children, currentPath);
          case 'unchanged':
            return [];
          default:
            throw new Error(`Unknown node type: ${node.type}`);
        }
      })
      .join('\n');

    return iter(diff);
  },

  stylish: (diff) => {
    const iter = (nodes, depth = 1) => {
      const indent = ' '.repeat(4 * depth - 2);
      return nodes.map((node) => {
        switch (node.type) {
          case 'added':
            return `${indent}+ ${node.key}: ${stringifyStylish(node.value, depth)}`;
          case 'removed':
            return `${indent}- ${node.key}: ${stringifyStylish(node.value, depth)}`;
          case 'updated':
            return [
              `${indent}- ${node.key}: ${stringifyStylish(node.oldValue, depth)}`,
              `${indent}+ ${node.key}: ${stringifyStylish(node.value, depth)}`,
            ].join('\n');
          case 'nested':
            return `${indent}  ${node.key}: {\n${iter(node.children, depth + 1)}\n${indent}  }`;
          case 'unchanged':
            return `${indent}  ${node.key}: ${stringifyStylish(node.value, depth)}`;
          default:
            throw new Error(`Unknown node type: ${node.type}`);
        }
      }).join('\n');
    };

    return `{\n${iter(diff)}\n}`;
  },

  json: (diff) => JSON.stringify(diff, null, 2),
};

const genDiff = (filepath1, filepath2, formatName = 'stylish') => {
  if (!formatters[formatName]) {
    throw new Error(`Unknown format: ${formatName}`);
  }

  const data1 = getParsedFile(filepath1);
  const data2 = getParsedFile(filepath2);

  const diff = buildDiff(data1, data2);

  return formatters[formatName](diff);
};

export default genDiff;
