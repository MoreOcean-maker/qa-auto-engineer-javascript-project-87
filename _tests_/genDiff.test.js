import { readFileSync } from 'fs';
import { fileURLToPath } from 'url';
import path from 'path';
import genDiff from '../src/genDiff.js';

const dirname = path.dirname(fileURLToPath(import.meta.url));
const getFixturePath = (filename) => path.join(dirname, '..', '_fixtures_', filename);

describe('genDiff', () => {
  test('correctly compares two JSON files', () => {
    const file1Path = getFixturePath('file1.json');
    const file2Path = getFixturePath('file2.json');
    const result = genDiff(file1Path, file2Path);
    expect(result).toMatchInlineSnapshot(`
      "{
          host: hexlet.io
        - timeout: 50
        + timeout: 20
        - proxy: 123.234.53.22
        - follow: false
        + verbose: true
      }"
    `);
  });
});
