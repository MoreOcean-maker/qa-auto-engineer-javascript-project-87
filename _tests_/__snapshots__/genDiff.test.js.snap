// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`genDiff json format 1`] = `
"[
  {
    "key": "follow",
    "type": "removed",
    "value": false
  },
  {
    "key": "host",
    "type": "unchanged",
    "value": "hexlet.io"
  },
  {
    "key": "proxy",
    "type": "removed",
    "value": "123.234.53.22"
  },
  {
    "key": "timeout",
    "type": "updated",
    "value": 20,
    "oldValue": 50
  },
  {
    "key": "verbose",
    "type": "added",
    "value": true
  }
]"
`;

exports[`genDiff plain format 1`] = `
"Property 'follow' was removed
Property 'proxy' was removed
Property 'timeout' was updated. From 20 to 50
Property 'verbose' was added with value: true"
`;

exports[`genDiff stylish format (default) 1`] = `
"{
  - follow: false
    host: hexlet.io
  - proxy: 123.234.53.22
  - timeout: 50
  + timeout: 20
  + verbose: true
}"
`;
